File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.cpp
class Solution {
public:
    int dp[52]={[0 ...51]=0};
//     int recur(string s,set<string>&set,int idx){
//         if(idx>=s.size())return 0;
//         int ans=0;
//         if(dp[idx]!=-1)return dp[idx];
//         ans=recur(s,set,idx+1);

//         for(int i=0;i+idx<s.size();i++){
//             if(set.find(s.substr(idx,i+1))!=set.end()){
//                 ans=max(i+1+recur(s,set,idx+i+1),ans);
                
//             }
           
//         }
//         return dp[idx]= ans;
//     }
    
    int minExtraChar(string s, vector<string>& dictionary) {
        set<string>set;
        for(string s:dictionary){
            set.insert(s);
        }
        // return s.size()- recur(s,set,0);
        int n=s.size();
        for(int i=n-1;i>=0;i--){
            dp[i]=dp[i+1];
            for(int j=1;i+j<=s.size();j++){
                if(set.find(s.substr(i,j))!=set.end()){
                    dp[i]=max(dp[i],j+dp[i+j]);
                }
            }
        }
        
        
        
        return n-dp[0];
    }
};

File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.java
class Solution {
    
    int []dp;
    Set<String>set;  
    public int memo(String s,int idx){
        int n=s.length();
        if(idx>=n)return 0;
        
        if(dp[idx]!=-1)return dp[idx];
    
        int ans=memo(s,idx+1);
        
        for(int i=1;i+idx<=n;i++)
        {
            if(set.contains(s.substring(idx,idx+i))){
                ans=Math.max(ans,i+memo(s,idx+i));
            }
        }
        return dp[idx]=ans;
        
    }
    
    public int minExtraChar(String s, String[] dictionary) {
        
        int n=s.length();
        dp=new int[n+1];
        Arrays.fill(dp,-1);
        set=new HashSet<>();
        for(String st:dictionary){
            set.add(st);
        }
        
        return n-memo(s,0);
        
    }
}